File: app/src/main/java/com/divercity/android/features/login/step1/LogInPageViewModel.java
package com.divercity.android.features.login.step1;

import android.app.Application;

import com.divercity.android.R;
import com.divercity.android.core.base.viewmodel.BaseViewModel;
import com.divercity.android.core.utils.SingleLiveEvent;
import com.divercity.android.core.utils.Util;
import com.divercity.android.data.Resource;
import com.divercity.android.data.networking.config.DisposableObserverWrapper;
import com.divercity.android.features.login.step1.usecase.ChecIskEmailRegisteredUseCase;
import com.divercity.android.features.login.step1.usecase.ConnectLinkedInApiHelper;
import com.divercity.android.features.login.step1.usecase.LoginFacebookUseCase;
import com.divercity.android.model.user.User;
import com.google.gson.JsonElement;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import androidx.lifecycle.MutableLiveData;
import io.reactivex.observers.DisposableObserver;

/**
 * Created by lucas on 26/09/2018.
 */

public class LogInPageViewModel extends BaseViewModel {

    private ChecIskEmailRegisteredUseCase checIskEmailRegisteredUseCase;
    private LoginFacebookUseCase loginFacebookUseCase;
    private Application application;
    private ConnectLinkedInApiHelper linkedInApiHelper;

    private SingleLiveEvent<Resource<Boolean>> isEmailRegistered = new SingleLiveEvent<>();
    private SingleLiveEvent<Resource<User>> loginFacebookResponse = new SingleLiveEvent<>();

    private SingleLiveEvent<Object> navigateToSignUp = new SingleLiveEvent<>();
    private SingleLiveEvent<Object> navigateToLogin = new SingleLiveEvent<>();

    @Inject
    public LogInPageViewModel(Application application,
                              ChecIskEmailRegisteredUseCase checIskEmailRegisteredUseCase,
                              ConnectLinkedInApiHelper linkedInApiHelper,
                              LoginFacebookUseCase loginFacebookUseCase) {
        this.application = application;
        this.loginFacebookUseCase = loginFacebookUseCase;
        this.checIskEmailRegisteredUseCase = checIskEmailRegisteredUseCase;
        this.linkedInApiHelper = linkedInApiHelper;
    }

    public void checkIfEmailRegistered(String email) {
        if (Util.isValidEmail(email.trim())) {
            isEmailRegistered.setValue(Resource.Companion.loading(null));
            DisposableObserver<Boolean> disposable = new DisposableObserver<Boolean>() {
                @Override
                public void onNext(Boolean r) {
                    isEmailRegistered.setValue(Resource.Companion.success(r));
                    if(r)
                        navigateToLogin.call();
                    else
                        navigateToSignUp.call();
                }

                @Override
                public void onError(Throwable e) {
                    isEmailRegistered.setValue(Resource.Companion.error(e.getMessage(),null));
                }

                @Override
                public void onComplete() {

                }
            };
            checIskEmailRegisteredUseCase.execute(disposable, ChecIskEmailRegisteredUseCase.Params.forCheckEmail(email));
        } else
            isEmailRegistered.setValue(Resource.Companion.error(application.getResources().getString(R.string.insert_valid_email), null));
    }

    public void loginFacebook(String token) {
        loginFacebookResponse.setValue(Resource.Companion.loading(null));

        DisposableObserverWrapper<User> disposable = new DisposableObserverWrapper<User>() {
            @Override
            protected void onFail(@NotNull String error) {
                loginFacebookResponse.setValue(Resource.Companion.error(error,null));
            }

            @Override
            protected void onHttpException(JsonElement error) {
                loginFacebookResponse.setValue(Resource.Companion.error(error.toString(),null));
            }

            @Override
            protected void onSuccess(@NotNull User loginResponse) {
                loginFacebookResponse.setValue(Resource.Companion.success(loginResponse));
            }
        };

        loginFacebookUseCase.execute(disposable, LoginFacebookUseCase.Params.forFacebook(token));
    }

    public void loginGoogle(String token) {
        // TODO
    }

    public MutableLiveData<Resource<Boolean>> getIsEmailRegistered() {
        return isEmailRegistered;
    }

    public SingleLiveEvent<Resource<User>> getLoginFacebookResponse() {
        return loginFacebookResponse;
    }

    public SingleLiveEvent<Object> getNavigateToSignUp() {
        return navigateToSignUp;
    }

    public SingleLiveEvent<Object> getNavigateToLogin() {
        return navigateToLogin;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        loginFacebookUseCase.dispose();
        checIskEmailRegisteredUseCase.dispose();
    }
}

File: app/src/main/java/com/divercity/android/features/login/step1/LogInPageViewModel.kt
package com.divercity.android.features.login.step1

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.divercity.android.R
import com.divercity.android.core.base.viewmodel.BaseViewModel
import com.divercity.android.core.utils.SingleLiveEvent
import com.divercity.android.core.utils.Util
import com.divercity.android.data.Resource
import com.divercity.android.data.Resource.Companion.error
import com.divercity.android.data.Resource.Companion.loading
import com.divercity.android.data.Resource.Companion.success
import com.divercity.android.data.networking.config.DisposableObserverWrapper
import com.divercity.android.features.login.step1.usecase.*
import com.divercity.android.model.user.User
import com.google.gson.JsonElement
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

/**
 * Created by lucas on 26/09/2018.
 */
class LogInPageViewModel @Inject constructor(
    private val application: Application,
    private val loginEmailUseCase: LoginEmailUseCase,
    private val requestResetPasswordUseCase: RequestResetPasswordUseCase,
    private val checkIskEmailRegisteredUseCase: CheckIskEmailRegisteredUseCase,
    private val loginFacebookUseCase: LoginFacebookUseCase,
    private val loginGoogleUseCase: LoginGoogleUseCase
) :
    BaseViewModel() {
    val checkEmailRegisteredResponse = SingleLiveEvent<Resource<Boolean>>()
    val loginEmailResponse = SingleLiveEvent<Resource<User>>()
    val requestResetPasswordResponse = SingleLiveEvent<Resource<Unit>>()
    val loginFacebookResponse = SingleLiveEvent<Resource<User>>()
    val loginGoogleResponse = SingleLiveEvent<Resource<User>>()
    val navigateToSignUp = SingleLiveEvent<Any>()
    val navigateToLogin = SingleLiveEvent<Any>()

    fun checkIfEmailRegistered(email: String) {
        if (Util.isValidEmail(email.trim { it <= ' ' })) {
            checkEmailRegisteredResponse.setValue(
                loading<Boolean>(
                    null
                )
            )
            val disposableObserver: DisposableObserver<Boolean> = object : DisposableObserver<Boolean>() {
                override fun onNext(r: Boolean) {
                    checkEmailRegisteredResponse.setValue(
                        success(
                            r
                        )
                    )
//                    if (r) navigateToLogin.call() else navigateToSignUp.call()
                }

                override fun onError(e: Throwable) {
                    checkEmailRegisteredResponse.setValue(
                        error<Boolean>(
                            e.message!!,
                            null
                        )
                    )
                }

                override fun onComplete() {}
            }
            checkIskEmailRegisteredUseCase.execute(
                disposableObserver,
                CheckIskEmailRegisteredUseCase.Params.forCheckEmail(email)
            )
        } else checkEmailRegisteredResponse.setValue(
            error<Boolean>(
                application.resources.getString(R.string.insert_valid_email),
                null
            )
        )
    }

    fun loginEmail(email: String?, password: String?) {
        if (password != null && password != "") {
            loginEmailResponse.value = loading<User>(null)
            val callback = object : LoginEmailUseCase.Callback() {
                override fun onFail(error: String) {
                    loginEmailResponse.value = error<User>(error, null)
                }

                override fun onSuccess(response: User) {
                    loginEmailResponse.value = success(response)
                }
            }
            loginEmailUseCase.execute(callback, LoginEmailUseCase.Params.forLogin(email, password))
        } else {
            loginEmailResponse.setValue(
                error<User>(
                    application.resources.getString(R.string.insert_valid_password),
                    null
                )
            )
        }
    }

    fun requestResetPassword(email: String) {
        requestResetPasswordResponse.value = loading(null)
        val callback = object : DisposableObserverWrapper<Unit>() {

            override fun onFail(error: String) {
                requestResetPasswordResponse.value = error(error, null)
            }

            override fun onHttpException(error: JsonElement) {
                requestResetPasswordResponse.value = error(error.toString(), null)
            }

            override fun onSuccess(t: Unit) {
                requestResetPasswordResponse.value = success(t)
            }
        }
        requestResetPasswordUseCase.execute(callback, RequestResetPasswordUseCase.Params.to(email))
    }

    fun loginFacebook(token: String?) {
        loginFacebookResponse.setValue(
            loading<User>(
                null
            )
        )
        val disposable: DisposableObserverWrapper<User> = object : DisposableObserverWrapper<User>() {
                override fun onFail(error: String) {
                    loginFacebookResponse.setValue(
                        error<User>(
                            error,
                            null
                        )
                    )
                }

                override fun onHttpException(error: JsonElement) {
                    loginFacebookResponse.setValue(
                        error<User>(
                            error.toString(),
                            null
                        )
                    )
                }

                override fun onSuccess(loginResponse: User) {
                    loginFacebookResponse.setValue(
                        success(
                            loginResponse
                        )
                    )
                }
            }
        loginFacebookUseCase.execute(disposable, LoginFacebookUseCase.Params.forFacebook(token))
    }

    fun loginGoogle(token: String?) {
        loginGoogleResponse.setValue(
            loading<User>(
                null
            )
        )
        val disposable: DisposableObserverWrapper<User> = object : DisposableObserverWrapper<User>() {
            override fun onFail(error: String) {
                loginGoogleResponse.setValue(
                    error<User>(
                        error,
                        null
                    )
                )
            }

            override fun onHttpException(error: JsonElement) {
                loginGoogleResponse.setValue(
                    error<User>(
                        error.toString(),
                        null
                    )
                )
            }

            override fun onSuccess(loginResponse: User) {
                loginGoogleResponse.setValue(
                    success(
                        loginResponse
                    )
                )
            }
        }
        loginGoogleUseCase.execute(disposable, LoginGoogleUseCase.Params.forGoogle(token))
    }

    fun getcheckEmailRegisteredResponse(): MutableLiveData<Resource<Boolean>> {
        return checkEmailRegisteredResponse
    }

    override fun onCleared() {
        super.onCleared()
        loginEmailUseCase.dispose()
        loginFacebookUseCase.dispose()
        loginGoogleUseCase.dispose()
        checkIskEmailRegisteredUseCase.dispose()
    }

}